Sources/StreamChat/AudioPlayer/AudioPlaying.swift

```diff
@@ -0,0 +1,358 @@
+            )
+        }
+    }
+
+    open func updateRate(_ newRate: AudioPlaybackRate) {
+        player.rate = newRate.rawValue
+    }
+
+    open func seek(to time: TimeInterval) {
+        player.pause()
+        updateContext { value in
+            value.currentTime = time
+            value.isSeeking = true
+        }
+        executeSeek(to: time)
+    }
+
+    // MARK: - Helpers
+
+    private func setUp() {
+        let player = self.player
+        let interval = CMTime(
+            seconds: 0.1,
+            preferredTimescale: CMTimeScale(NSEC_PER_SEC)
+        )
+
+        playerObserver.addPeriodicTimeObserver(
+            player,
+            forInterval: interval,
+            queue: nil
+        ) { [weak self] in
+            guard let self = self, self.context.isSeeking == false else {
+                return
+            }
+
+            self.updateContext { value in
+                let currentTime = player.currentTime().seconds
+                value.currentTime = currentTime.isFinite && !currentTime.isNaN
+                    ? TimeInterval(currentTime)
+                    : .zero
+
+                value.isSeeking = false
+
+                value.state = player.rate != 0 ? .playing : .paused
+
+                value.rate = .init(rawValue: player.rate)
+            }
+        }
+
+        playerObserver.addTimeControlStatusObserver(
+            player
+        ) { [weak self] newValue in
+            guard let self = self, let newValue = newValue else {
+                return
+            }
+
+            let currentPlaybackState = self.context.state
+
+            self.updateContext { value in
+                switch (newValue, currentPlaybackState) {
+                case (.playing, .playing), (.paused, .paused):
+                    break
+                case (.paused, .playing), (.paused, .loading):
+                    value.state = .paused
+                    value.rate = .zero
+                case (.playing, .paused), (.playing, .stopped), (.playing, .loading):
+                    value.state = .playing
+                default:
+                    log.debug("\(type(of: self)): No action for transition \(currentPlaybackState) -> \(newValue)", subsystems: .audioPlayback)
+                }
+            }
+        }
+
+        playerObserver.addStoppedPlaybackObserver(
+            queue: nil
+        ) { [weak self] playerItem in
+            guard
+                let self = self,
+                let playerItemURL = (playerItem.asset as? AVURLAsset)?.url,
+                let currentItemURL = (player.currentItem?.asset as? AVURLAsset)?.url,
+                playerItemURL == currentItemURL
+            else {
+                return
+            }
+            var newContext = self.context
+            newContext.state = .stopped
+            newContext.currentTime = 0
+            newContext.rate = .zero
+            newContext.isSeeking = false
+            self.context = newContext
```

Comment:
Shouldn't this use the `updatedContext` function?

Reply:
Yes it should!